### PR TITLE
[SOL-2137] Add Backend Validation for Coupon Creation - Benefit Value

### PR DESCRIPTION
**Pull request by iivic**
Values for benefit value that are not allowed: string, empty, negative numbers, when benefit type 'Percentage' max value is 100. https://openedx.atlassian.net/browse/SOL-2137
@vkaracic @mjfrey @marjev  Please review.

### CODE REVIEW

**Comment by openedx-webhooks** [comment]
Thanks for the pull request, @iivic! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request.

<a   href="https://openedx-webhooks.herokuapp.com/github/process_pr?repo=edx%2Fecommerce&number=1017"   rel="nofollow" >Create an OSPR issue for this pull request</a>.

File: ecommerce/extensions/offer/models.py [hunk 1]
```diff
@@ -1,4 +1,5 @@
 # noinspection PyUnresolvedReferences
+import logging
```
Comment (vkaracic) [inline ecommerce/extensions/offer/models.py]: Alphabetize this please. :)

File: ecommerce/extensions/offer/models.py [hunk 2]
```diff
@@ -21,12 +22,23 @@ def save(self, *args, **kwargs):
...
 
     def clean_type(self):
         if self.type not in VALID_BENEFIT_TYPES:
+            logger.exception(
+                'Failed to create Benefit. Benefit type must be one of the following %s.', VALID_BENEFIT_TYPES
+            )
             raise ValidationError(_('Unrecognised benefit type {type}'.format(type=self.type)))
 
+    def clean_value(self):
+        if self.value < 0:
```
Comment (vkaracic) [inline ecommerce/extensions/offer/models.py:36]: And what happens if it's > 100?

File: ecommerce/extensions/offer/models.py [hunk 3]
```diff
@@ -21,12 +22,23 @@ def save(self, *args, **kwargs):
...
+            logger.exception(
+                'Failed to create Benefit. Benefit value may not be a negative number.'
```
Comment (vkaracic) [inline ecommerce/extensions/offer/models.py]: This and the logger message above can fit in one line.

**Review by vkaracic — CHANGES_REQUESTED** [review]

File: ecommerce/extensions/offer/models.py (hunk 2 above)
Comment (iivic) [reply ecommerce/extensions/offer/models.py:36]: Oscar validates this case if the type is Percentage oscar raises a ValidationError if the type is 'Absolute' nothing happens.

**Comment by iivic** [comment]
@clintonb

**Review by vkaracic — APPROVED** [review]
LGTM

**Review by clintonb — APPROVED** [review]
Remember to squash your commits.

File: ecommerce/extensions/api/v2/tests/views/test_coupons.py [hunk 4]
```diff
@@ -262,21 +262,29 @@ def test_create_serializer_data(self):
         self.assertEqual(details_response['coupon_type'], 'Enrollment code')
         self.assertEqual(details_response['code_status'], 'ACTIVE')
 
-    @ddt.data(
-        ({'category': {'a': 'a', 'b': 'b'}}, status.HTTP_400_BAD_REQUEST),
```
Comment (marjev) [inline ecommerce/extensions/api/v2/tests/views/test_coupons.py:266]: Why is this removed? This should still fail and is there as something that can happen and should return at least 400 BAD REQUEST status or something better.

**Review by marjev — CHANGES_REQUESTED** [review]

File: ecommerce/extensions/api/v2/tests/views/test_coupons.py (hunk 4 above)
Comment (iivic) [reply ecommerce/extensions/api/v2/tests/views/test_coupons.py:266]: It's not removed I moved it to `test_create_coupon_product_invalid_data` test method.

**Review by marjev — APPROVED** [review]

**Pull request merged**